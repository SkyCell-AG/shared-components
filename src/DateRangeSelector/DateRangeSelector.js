import React from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'

import DateSelect from 'DateSelect'
import rangeToDayPassed from 'utils/rangeToDayPassed'
import {
    strToDate,
} from 'utils/DateUtils'

import useStyles from './DateRangeSelector.style'

const propTypes = {
    mini: PropTypes.bool,
    options: PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string.isRequired,
        value: PropTypes.number.isRequired,
    })),
    selectOption: PropTypes.func.isRequired,
    onChangeRange: PropTypes.func.isRequired,
    showTimeRange: PropTypes.bool,
    from: PropTypes.string.isRequired,
    to: PropTypes.string.isRequired,
}

const defaultProps = {
    mini: false,
    showTimeRange: false,
    options: [
        {
            label: '24h',
            value: 1,
        },
        {
            label: '7d',
            value: 7,
        },
        {
            label: '14d',
            value: 14,
        },
    ],
}

const DateRangeSelector = ({
    mini,
    options,
    selectOption,
    showTimeRange,
    onChangeRange,
    from,
    to,
}) => {
    const classes = useStyles()
    const dayPassed = from && to ? rangeToDayPassed({
        from: strToDate(from),
        to: strToDate(to),
    }) : undefined

    return (
        <div className={classes.wrapper}>
            {!mini && (
                <div className={classes.optionsWrapper}>
                    {options.map(({
                        label,
                        value,
                    }) => {
                        return (
                            <div
                                data-testid={`date-range-button-${value}`}
                                key={value}
                                className={clsx(classes.option, {
                                    [classes.selectedOption]: dayPassed === value,
                                })}
                                onMouseDown={() => {
                                    selectOption(value)
                                }}
                            >
                                {label}
                            </div>
                        )
                    })}
                </div>
            )}
            {showTimeRange && (
                <div className={classes.inputsWrapper}>
                    <div className={classes.input}>
                        <DateSelect
                            title="From"
                            className={classes.datePicker}
                            onChange={onChangeRange}
                            name="from"
                            value={from}

                        />
                    </div>
                    <div className={classes.input}>
                        <DateSelect
                            title="To"
                            className={classes.datePicker}
                            onChange={onChangeRange}
                            name="to"
                            value={to}
                        />
                    </div>
                </div>
            )}
        </div>
    )
}

DateRangeSelector.propTypes = propTypes
DateRangeSelector.defaultProps = defaultProps

export default DateRangeSelector
