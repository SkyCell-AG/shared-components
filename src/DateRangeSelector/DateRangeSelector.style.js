import {
    makeStyles,
} from '@mui/styles'
import {
    useTheme,
} from '@mui/material/styles'

const useStyles = makeStyles(() => {
    const theme = useTheme()

    return {
        wrapper: {
            display: 'flex',
            justifyContent: 'space-between',
            fontSize: theme.typography.h4.fontSize,
        },

        optionsWrapper: {
            display: 'flex',
            alignItems: 'center',
        },

        option: {
            cursor: 'pointer',
            padding: theme.spacing(0.625),
            width: 40,
            textAlign: 'center',
            transition: 'background-color 03:s',
            '&:hover': {
                backgroundColor: theme.palette.secondary[100],
            },
        },

        selectedOption: {
            backgroundColor: `${theme.palette.secondary[100]} !important`,
        },

        inputsWrapper: {
            display: 'flex',
        },

        input: {
            alignItems: 'center',
            marginLeft: theme.spacing(2.5),
        },

        dateContainer: {
            display: 'flex',
            fontSize: '1.2rem',
        },

        editedDate: {
            color: 'silver',
        },

        dayPickerContainer: {
            display: 'none',
            position: 'absolute',
            width: '550px',
            marginLeft: '-310px',
        },

        dayPickerOpened: {
            display: 'block',
        },

        datePicker: {
            marginLeft: theme.spacing(1.25),
        },
    }
})

export default useStyles
