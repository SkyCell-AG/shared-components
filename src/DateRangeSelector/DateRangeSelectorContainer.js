import React, {
    useCallback,
} from 'react'
import PropTypes from 'prop-types'
import noop from 'lodash/noop'

import {
    dayPassedToRange,
    dateToDateStr,
} from 'utils/DateUtils'

import DateRangeSelector from './DateRangeSelector'

const propTypes = {
    value: PropTypes.shape({
        from: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        to: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
    }).isRequired,
    onChange: PropTypes.func,
    setDateRange: PropTypes.func.isRequired,
}

const defaultProps = {
    onChange: noop,
}

const DateRangeSelectorContainer = ({
    value = {},
    onChange,
    setDateRange,
    ...rest
}) => {
    const {
        from,
        to,
    } = value

    const selectOption = useCallback((selected) => {
        onChange(dayPassedToRange(selected))
        setDateRange(false)
    }, [
        onChange,
        setDateRange,
    ])

    const onChangeRange = useCallback((_, {
        target: {
            name,
            value: newValue,
        },
    }) => {
        const updatedData = {
            from: new Date(from).setHours(0, 0, 0),
            to: new Date(to).setHours(23, 59, 59),
            [name]: newValue,
        }

        onChange(updatedData, {
            name,
            value: updatedData,
        })

        setDateRange(true)
    }, [
        from,
        onChange,
        setDateRange,
        to,
    ])

    return (
        <DateRangeSelector
            {...rest}
            selectOption={selectOption}
            onChangeRange={onChangeRange}
            from={dateToDateStr(from)}
            to={dateToDateStr(to)}
        />
    )
}

DateRangeSelectorContainer.propTypes = propTypes
DateRangeSelectorContainer.defaultProps = defaultProps

export default DateRangeSelectorContainer
