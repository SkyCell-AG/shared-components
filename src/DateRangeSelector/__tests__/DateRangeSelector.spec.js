import React from 'react'
import {
    shallow,
    mount,
} from 'enzyme'

import DateRangeSelector from '../DateRangeSelector'

describe('shared-components/DateRangeSelector', () => {
    it('should match snapshot', () => {
        const wrapper = shallow(
            <DateRangeSelector
                options={[
                    {
                        label: '24h',
                        value: 1,
                    },
                    {
                        label: '7d',
                        value: 7,
                    },
                    {
                        label: '14d',
                        value: 14,
                    },
                ]}
                selectOption={jest.fn()}
                onChangeRange={jest.fn()}
                from="16.08.2020"
                to="18.08.2020"
                setDateRange={jest.fn()}
            />,
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('should show time range', () => {
        const wrapper = shallow(
            <DateRangeSelector
                options={[
                    {
                        label: '24h',
                        value: 1,
                    },
                    {
                        label: '7d',
                        value: 7,
                    },
                    {
                        label: '14d',
                        value: 14,
                    },
                ]}
                selectOption={jest.fn()}
                showTimeRange
                onChangeRange={jest.fn()}
                from="16.08.2020"
                to="18.08.2020"
                setDateRange={jest.fn()}
            />,
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('renders without crashing', () => {
        mount(
            <DateRangeSelector
                selectOption={jest.fn()}
                onChangeRange={jest.fn()}
                from="16.08.2020"
                to="18.08.2020"
                setDateRange={jest.fn()}
            />,
        )
    })
})
