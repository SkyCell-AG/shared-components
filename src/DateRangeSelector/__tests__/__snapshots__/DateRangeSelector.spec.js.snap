// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared-components/DateRangeSelector should match snapshot 1`] = `
<div
  className="makeStyles-wrapper-1"
>
  <div
    className="makeStyles-optionsWrapper-2"
  >
    <div
      className="makeStyles-option-3"
      key="1"
      onMouseDown={[Function]}
    >
      24h
    </div>
    <div
      className="makeStyles-option-3"
      key="7"
      onMouseDown={[Function]}
    >
      7d
    </div>
    <div
      className="makeStyles-option-3"
      key="14"
      onMouseDown={[Function]}
    >
      14d
    </div>
  </div>
</div>
`;

exports[`shared-components/DateRangeSelector should show time range 1`] = `
<div
  className="makeStyles-wrapper-1"
>
  <div
    className="makeStyles-optionsWrapper-2"
  >
    <div
      className="makeStyles-option-3"
      key="1"
      onMouseDown={[Function]}
    >
      24h
    </div>
    <div
      className="makeStyles-option-3"
      key="7"
      onMouseDown={[Function]}
    >
      7d
    </div>
    <div
      className="makeStyles-option-3"
      key="14"
      onMouseDown={[Function]}
    >
      14d
    </div>
  </div>
  <div
    className="makeStyles-inputsWrapper-5"
  >
    <div
      className="makeStyles-input-6"
    >
      <DateSelectContainer
        className="makeStyles-datePicker-7"
        name="from"
        onChange={[MockFunction]}
        title="From"
        value="16.08.2020"
      />
    </div>
    <div
      className="makeStyles-input-6"
    >
      <DateSelectContainer
        className="makeStyles-datePicker-7"
        name="to"
        onChange={[MockFunction]}
        title="To"
        value="18.08.2020"
      />
    </div>
  </div>
</div>
`;
